### ISSUE TITLE
Validation logging frequency of `TensorBoardLogger` is different than expected and has the same loss values in step and epoch

### ISSUE THREAD

**Issue by nicoloesch** [Lightning-AI/lightning#13525]
## 🐛 Bug

The validation logging frequency is different from the one in the training for the `TensorBoardLogger`. Additionally, default logger logs the exact same values in the `validation_step` as in the `validation_epoch` for the `loss`.
Overwriting `on_epoch` and `on_step` to `True` results in a log **every** step for the validation instead of every `log_every_n_steps`.

### To Reproduce

My dataloader is defined as follows for **both validation and training**:

- `num_subjects`: 10
- `num_patches_per_subject`: 10 ( = 100 patches per epoch)
- `batch_size`: 2 (= 50 steps per epoch)

Other logging-related configurations:

- `max_epochs`: 5 (to get some data points)
- Logger: `TensorBoardLogger` (default)
- `log_every_n_steps`: 25 (to see a difference between `epoch` and `step` in the logger as I am expecting to log twice per epoch for the steps [50/25 = 2]

The logging call is defined as follows: 

```python

self.log(f"{stage}_{step_or_epoch}_{metric_name}", dict_to_log,  batch_size=self.hparams.batch_size)

```

with `stage` in `['val', 'train']` as I unified both in a single function for the steps called `_step` as they are similar in what they do, `step_or_epoch` in `['step', 'epoch']` and a `dict_to_log` for each of the metrics I am logging (accuracy, dice, iou and hd [all of them expect the HD being of `torchmetrics` origin however with the functional interface currently]). I am logging a dictionary, as all of the metrics are then contained in a single plot in the tensorboard python viewer. The `dict_to_log` looks like the following:

```python

dict_to_log = {
    'class_1': torch.Tensor,
    'class_2': torch.Tensor,
    'class_3': torch.Tensor,
}

```
with each `torch.Tensor` representing a single value for the respective class. In the case of a `step`, I average over the batch size to get the single tensor. In the `epoch`, I am concatenating all returns of the `<stage>_step` for each metric and then calculate the mean over it to get the single tensor.

### Expected behavior

If I run the model for the specified 5 epochs, I would expect the following for **both validation and training logs**:

- \# of step data points: `10  = (5 x 2)` as I am logging twice per epoch for 5 epochs
- \# of epoch data points: `5` as I am logging once at the end of the epoch.

**However, this only works in the training.** The `validation_step` and `validation_epoch` contain both only 5 entries., with the entries showing the exact same values for the `loss` but not for other metrics.

### Environment

* CUDA:
	- GPU:
		- NVIDIA GeForce MX450
	- available:         True
	- version:           11.3
* Packages:
	- numpy:             1.22.3
	- pyTorch_debug:     False
	- pyTorch_version:   1.11.0
	- pytorch-lightning: 1.6.4
	- tqdm:              4.64.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.9.12
	- version:           \# 43-Ubuntu SMP Wed Jun 15 12:54:21 UTC 2022
* IDE:
	- PyCharm 2022.1.2

### Additional context

I only observed this while debugging an error with my model, which showed overfitting. However, both step and epoch of the `validation_loss` showed the exact same graph and also logged the exact same values upon investigation.

Overwriting `on_epoch` and `on_step` to `True` in `self.log()` did result in logging the `validation_step` literally every step, resulting in 250 logs.

### Potential Workaround

Only call `self.log()`, if I reached the `log_every_n_steps`.

cc @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy @akihironitta

**Comment by nicoloesch** [comment]
### Update (some time later after trying several combinations and trying my own Logger)

Instead of using the `self.log()` method, I create now a `Dict[str, Union[float, Dict[str, float]]]`, that contains all the strings of the metrics I want to log as the keys and the respective either single tensor or the dictionary containing single tensors as the vector.
That dictionary is then directly passed to `self.logger.log_metrics(dict_to_log, step)`. As a result, I am getting my desired output lengths of the log. 

It seems like, the `self.log()` method for the validation logs is either not correctly used by me or the internal logging/ automatic decision if one wants to log something is not working as I expect it to work with either too many (manual overwrite of `on_step` and `on_epoch`) or too few logs (automatic behavior).

I am however unsure, if my temporary "fix" is safe for the usage with multiple GPUs (my assumption is a big NO).
I therefore appreciate any comment you guys have regarding my issue! 

Cheers,
Nico

**Comment by carmocca** [comment]
It would be useful if you could share a simple script that has your configuration and asserts what is expected, showing the failure precisely. You can adapt this script: https://github.com/Lightning-AI/lightning/blob/master/examples/pytorch/bug_report/bug_report_model.py

**Comment by nicoloesch** [comment]
Hi @carmocca,

I recently switched now fully to the `TorchMetrics` implementations and their entire automatic calculation. I initially only had the above example to manually rename the metrics by storing the respective values in a dicionary after calling `compute` on each of the metrics. Therefore, I think this has been somewhat resolved, as I have not observed the behaviour with the full utilisation with `self.log` and the `TorchMetrics`.
